### ISSUE TITLE
Bind group inheritance

### ISSUE THREAD

**Issue by kvark** [gpuweb/gpuweb#171]
The topic about inheriting bind group state between pipelines has been brought up in various occasions but we haven't established a clear story for it:
  - https://github.com/gpuweb/gpuweb/pull/47#discussion_r168324254
  - "Inheritance" in https://github.com/gpuweb/gpuweb/issues/19
  - https://github.com/gpuweb/gpuweb/issues/167#issuecomment-454977897

We need a proper specification of how inheritance works as well as the notes on how the implementation can do it. I would very much prefer if, for the implementation, we don't require a check on every `draw()` call and instead do the appropriate checks on binding the pipelines and bind groups.

**Comment by magcius** [comment]
As an engine programmer, I do not mind if I have to re-bind after switching pipelines. Bind group inheritance seems mostly suited towards demos and tiny shortcuts -- any substantial codebase is going to have a state tracker to be able to rebind things when it needs to.

**Comment by grorg** [comment]
Discussed at [18 Mar 2019](https://docs.google.com/document/d/16kKh2JGE-tgOBx61qZUAAB0x4u8rSJus-aSuSyib35Q/) teleconference

**Comment by kvark** [comment]
Things from the call: more options.

4) invalidate the bind groups starting from the first incompatible one and up, upon binding a different layout - requires specifying the layout compatibility rules, aka Vulkan
5) don't invalidate anything! Semantically, match the bind groups with active pipeline layout at draw call site.

During the call, I suggested 5) to the group as the way to go. Main motivation is:
  - matching the semantics of vertex buffer bindings
  - avoiding surprises for the users that their bindings are invalidated

Implementation-wise, different strategies are possible in order to avoid full validation cost on every draw call. Dawn can figure out the validation spots during client-side recording and then use that information to properly validate on the server side and insert necessary binding code in native command buffers. wgpu-rs can eagerly bind resources when they are bound, remembering slots that don't have matching binding, to be recalled at draw call site.

**Comment by kvark** [comment]
TL;DR: the group decision so far is to go with the 5) approach with a note that bind group layouts are provided with "by value" semantics: two bind group layouts created from the same contents of the descriptor are considered equal.

**Comment by kainino0x** [comment]
Does that actually change the observable behavior, or is it just an implementation detail?

**Comment by magcius** [comment]
This is observable. If one does (in pseudocode):

```js
const bg1 = device.createBindGroupLayout({ descriptors: [{ type: 'uniform-buffer' }] });
const bg2 = device.createBindGroupLayout({ descriptors: [{ type: 'uniform-buffer' }] });

const pipeline1 = device.createPipeline({ bindGroupLayout: bg1, ... });
const pipeline2 = device.createPipeline({ bindGroupLayout: bg2, ... });

const bindings = device.createBindings({ bindGroupLayout: bg1, ... });

passEncoder.setPipeline(pipeline1);
passEncoder.setBindings(bindings);
passEncoder.setPipeline(pipeline2);
// No need to call passEncoder.setBindings(bindings);
```

bg1 and bg2 are compatible, so bind-group inheritance applies, and there's no need to rebind. I'm not sure *why* that's important to be in the spec, but it's not that hard for a backend to do and it sounds like backends will need to cache anyway to meet perf goals.

**Comment by kainino0x** [comment]
Option 5 is observable, but I meant to ask about the "by value" semantics / value equality. I don't know how that would be observed.

**Comment by kvark** [comment]
@kainino0x a program that is valid with by-ref semantics would also be valid with by-value semantics of the bind group layouts, but not the other way around. So the effect to observe is a validation error for an app that relies on de-duplication by the backend.

This makes this change one-directional: there is going to be no way back without breaking programs. At the same time, if we decide for by-ref semantics now, we can listen to feedback and flip this switch later without any consequences (part of the appeal to me).

**Comment by magcius** [comment]
I assume "by-value" meant that if I create two bind groups layouts from different objects that have the same value (like my example), I will get back two separate platform objects that are opaque, but still yet "by-value" compatible.

**Comment by kainino0x** [comment]
Sorry for not understanding, I don't see what the validation error would be. E.g. @magcius's example uses two different BGLs - wouldn't it pass validation under both by-value and by-ref?

**Comment by kvark** [comment]
@magcius right. I'm glad to hear another opinion of this feature not being needed.
@kainino0x the example would fail "by-ref" validation but pass with "by-value" semantics (since the dictionaries passed in `createBindGroupLayout` have the same contents).

**Comment by kainino0x** [comment]
Ok, I see. I thought the proposal was slightly more magic wrt inheritance, but I think I get how it works now.

In by-value, is there still value in compiling the BGLs? Just for apps to be able to compile them early?

**Comment by magcius** [comment]
I'm not sure what "compiling the BGLs" means.

**Comment by kvark** [comment]
@kainino0x sure, it still makes sense to actually create BGLs on the backend as early as possible. I understand @Kangz preferred solution as:
- there is a map somewhere that associates the BGL contents to actual native BGL objects of the backend
- when a new one is requested to be created, we check if it's already in the map and return early without doing any work
- other wise, the backend creates a new one at that point. No need to delay this till later.

**Comment by magcius** [comment]
To me, the argument for it being by-value is that on backends where you already have by-value semantics (which, admittedly, should likely be all of them), it might be harder to add checks for the higher-level object wrapper objects. And if the backend implementation checks by-value accidentally, it would create compatibility issues.

I'm fine with either. From a user perspective, by-value is not necessary to me, but from an implementation and compatibility perspective, it seems like a net win.

**Comment by kvark** [comment]
@magcius thanks for clarifications! Minor correction:
> And if the backend implementation checks by-value accidentally, it would create compatibility issues.

It's meant to be caught by validation, which should be happening before anything reaches a backend anyway.

**Comment by magcius** [comment]
Fair enough -- I don't know enough about implementations to know if validation will be shared between browsers, on which side of the network protocol fence they'll go on, etc., where the resource caching deduplication will happen, etc.

**Comment by kainino0x** [comment]
By compiling BGLs I just meant having a createBindGroupLayout, rather than just passing a GPUBindGroupLayoutDescriptor everywhere we currently have GPUBindGroupLayout today.

Validation won't be shared between browsers, it happens on the GPU process (because the content process is untrusted), and the deduplication would happen either next to validation or in the backend.

**Comment by magcius** [comment]
If we have by-value semantics then there's indeed less sense in making BGLs platform objects -- they could be deferred until createPipeline time. I'm not sure what the semantics of this are for different drivers. BGLs are strictly a Vulkan concept -- Metal and D3D12 have different binding models, as has been documented in excruciating detail elsewhere. I know that Vulkan made BGLs platform objects (known as VkDescriptorSetLayout), but I don't have visibility into which vendors were pushing for that. My guess is that the desktop drivers cannot do anything with the platform objects until vkCreatePipeline time.

**Comment by kainino0x** [comment]
I'd guess they're just a way for the driver to precompute the in-memory layout, and do not correspond to any on-device object. It aligns with Vulkan design to allow precomputation even of small things like layout so it can be reused, avoiding adding a cache inside the driver.

**Comment by magcius** [comment]
Most drivers I imagine want to do full-PSO optimization which is why I'm skeptical of the design. I guess having BGLs as a platform object lets them clarify inheritance rules much easier -- from my reading of the spec they definitely use by-ref inheritance semantics.

**Comment by kvark** [comment]
BGLs are semantically not Vulkan-only concept. They are the contract between bindings and pipelines. Both Vulkan and D3D12 conceptually encourage bindings scope to be larger than pipeline scope, hence the concepts of pipeline layout and root signature. On Metal it's not required but still beneficial (to reduce the amount of re-binding), especially when argument buffers are involved.

I'm really not a fan of turning objects (that could be created ahead of time on the backend) into dictionaries. And that's my main concern with by-value semantics - it encourages the path of dropping baked objects instead of dictionaries, the path of doing the work later and lazily.

If we want to go this way, the first thing to consider for removal would be `GPUPipelineLayout`: we could provide a list of BGLs when creating pipelines. In WebGPU that's the only place where we currently use it anyway, unlike Vulkan where the user has to provide it when binding descriptor sets and push constants.

**Comment by Kangz** [comment]
@magcius for bindless GPUs a VkDescriptorSetLayout is like a C-structure declaration and allows precomputing offsets like @kainino0x said. Also it is valid for a VkDescriptorSetLayout (or BGL) to have holes, and the driver or WebGPU implementations can pack the layouts the way they want.

In our current implementation BGL creation isn't super expensive, but BG creation will be one of the hot paths of the API so any amount of precomputation helps.

@kvark agreed that we could drop the pipeline layout in that case. Vulkan requires it for some commands because it doesn't want to have to keep state in the command buffers to know what's the current pipeline's layout.

I'm not sure I understand what drawbacks you see to by-value semantics, it makes sense that a bindgroup should be compatible with a pipeline using the same layout, because what matter is the layout, not the API object that encapsulates it. Having by-ref semantics could mean applications have to pipe the BGL between say a material component that creates render pipelines, and a model loading component that loads resources. Instead of just having conventions by-ref semantics force the BGL to be moved from one to the other which isn't very practical (and not needed in any of the APIs we target).

**Comment by magcius** [comment]
For any pipeline with more bindings than has room for in some cache (which is pretty often), a good driver would optimize the layout of the bindings based on what the shaders in the pipeline use -- it's very common to have unused bindings.

> Having by-ref semantics could mean applications have to pipe the BGL between say a material component that creates render pipelines, and a model loading component that loads resources.

So two independent components are relying on pipeline inheritance to save a single setBindings call, but it's too inconvenient to pass a shared variable through? That seems unrealistic to me.

In practice, in most engines, there is going to be a higher-level draw call structure which is iterated over and called immediately. You can see this in my renderer ( https://github.com/magcius/noclip.website/blob/master/src/gfx/render/GfxRenderer.ts#L473-L489 ), in Unreal's new 4.22 renderer ( https://youtu.be/qx1c190aGhs?t=740 ), and in other game codebases I have worked on but cannot point you at the source code for.

Since user code cannot themselves compare the platform objects for by-value equality, since it is an opaque object to the user, we cannot use by-value pipeline inheritance to our advantage, unless we implement a higher-level app cache in our own code. Hence why I say by-value semantics do not really matter for any real-world codebase. The only thing it is a guarantee of is some degree of platform caching, which I think is a good idea.

**Comment by kvark** [comment]
> For any pipeline with more bindings than has room for in some cache (which is pretty often), a good driver would optimize the layout of the bindings based on what the shaders in the pipeline use -- it's very common to have unused bindings.

This doesn't make sense to me. The user binds resources to the pipeline layout, then at some point switches the pipelines to draw them. If the driver does pipeline-specific optimizations of a layout, it would have to rebind resources either on pipeline change or a draw call, both of which are hotspots potentially.

> So two independent components are relying on pipeline inheritance to save a single setBindings call, but it's too inconvenient to pass a shared variable through? That seems unrealistic to me.

That I fully agree with. What's the point of encouraging copy-pasting of BGL dictionaries between different parts of someone's multi-layered software stack? A change in one place would then not manifest in an actual failure until a specific run-time combination comes into play. That seems significantly worse than failing at BG creation, which is where the error would happen with by-ref semantics of BGLs (and sharing of BGL handles between those layers of software that want to share them).

> The only thing it is is a guarantee of is some degree of platform caching, which I think is a good idea.

I don't know what savings are there for random sharing of BGLs. I assume these are really lightweight objects anyway, and there aren't many of them in an application.

**Comment by magcius** [comment]
> This doesn't make sense to me. The user binds resources to the pipeline layout, then at some point switches the pipelines to draw them. If the driver does pipeline-specific optimizations of a layout, it would have to rebind resources either on pipeline change or a draw call, both of which are hotspots potentially.

I cannot explain the intricacies of GPU architecture and drivers in this thread, but for a heavily simplfiied example -- if you have two tiers of memory (fast and slow) and there are commands to move between them for some fixed cost, it might be cheaper to eat that cost once if you know that it will speed up the shader that might otherwise read from the slow memory. The driver's responsibility is to be an optimizing compiler, to evaluate those performance tradeoffs and make the right call for all of them. The more context it has to make those decisions, the better.

> I don't know what savings are there for random sharing of BGLs. I assume these are really lightweight objects anyway, and there aren't many of them in an application.

I agree. My assumption was that there would be other places in the API to make the by-value / by-ref choice, and for the sake of consistency we would choose the by-value one. But I perhaps should have stated this out loud.

**Comment by Kangz** [comment]
This has been resolved in a WebGPU call: WebGPU will not have a notion of inheritance and simply at draw time the layout of the bindgroups required by the pipeline must match the layout of the currently set bindgroups at the same index.
